boot.php

```diff
@@ -781,60 +781,84 @@ function get_scheme() {
 		return($this->scheme);
 	}
 
+	/**
+	 * @brief Retrieves the Friendica instance base URL
+	 *
+	 * Caches both SSL and non-SSL version for performance
+	 *
+	 * @param bool $ssl
+	 * @return string
+	 */
 	function get_baseurl($ssl = false) {
 
 		// Is the function called statically?
-		if (!is_object($this))
+		if (!is_object($this)) {
 			return(self::$a->get_baseurl($ssl));
+		}
 
-		$scheme = $this->scheme;
+		if (!isset($this->baseurl[$ssl ? 'https' : 'http'])) {
+			$scheme = $this->scheme;
 
-		if((x($this->config,'system')) && (x($this->config['system'],'ssl_policy'))) {
-			if(intval($this->config['system']['ssl_policy']) === intval(SSL_POLICY_FULL))
-				$scheme = 'https';
+			if((x($this->config, 'system')) && (x($this->config['system'], 'ssl_policy'))) {
+				if(intval($this->config['system']['ssl_policy']) === intval(SSL_POLICY_FULL))
+					$scheme = 'https';
 
-			//	Basically, we have $ssl = true on any links which can only be seen by a logged in user
-			//	(and also the login link). Anything seen by an outsider will have it turned off.
+				//	Basically, we have $ssl = true on any links which can only be seen by a logged in user
+				//	(and also the login link). Anything seen by an outsider will have it turned off.
 
-			if($this->config['system']['ssl_policy'] == SSL_POLICY_SELFSIGN) {
-				if($ssl)
-					$scheme = 'https';
-				else
-					$scheme = 'http';
+				if($this->config['system']['ssl_policy'] == SSL_POLICY_SELFSIGN) {
+					if($ssl) {
+						$scheme = 'https';
+					} else {
+						$scheme = 'http';
+					}
+				}
 			}
-		}
 
-		if (get_config('config','hostname') != "")
-			$this->hostname = get_config('config','hostname');
+			if (get_config('config','hostname') != '') {
+				$this->hostname = get_config('config', 'hostname');
+			}
 
-		$this->baseurl = $scheme . "://" . $this->hostname . ((isset($this->path) && strlen($this->path)) ? '/' . $this->path : '' );
-		return $this->baseurl;
+			$this->baseurl[$ssl ? 'https' : 'http'] = $scheme . "://" . $this->hostname . ((isset($this->path) && strlen($this->path)) ? '/' . $this->path : '' );
+		}
+		return $this->baseurl[$ssl ? 'https' : 'http'];
```

Comment:
As a rule I try to avoid having multiple return statements in a single functions. It's generally easier to debug and when the function needs to stop prematurely, I prefer to use Exception.

But I definitely see your point. And I would tend to do it to avoid too many nested conditions, but I don't think it's warranted here.